packages/ozone/src/daemon/blob-diverter.ts

```diff
@@ -0,0 +1,229 @@
+import { dbLogger } from '../logger'
+import { BlobReportServiceConfig } from '../config'
+
+type PollState = {
+  timer?: NodeJS.Timer
+  promise: Promise<void>
+}
+
+export class BlobDiverter {
+  destroyed = false
+
+  pollState: PollState = {
+    promise: Promise.resolve(),
+  }
+
+  serviceConfig: BlobReportServiceConfig
+  idResolver: IdResolver
+
+  constructor(
+    public db: Database,
+    services: {
+      idResolver: IdResolver
+      serviceConfig: BlobReportServiceConfig
+    },
+  ) {
+    this.serviceConfig = services.serviceConfig
+    this.idResolver = services.idResolver
+  }
+
+  start() {
+    this.poll(this.pollState, () => this.divertBlob())
+  }
+
+  poll(state: PollState, fn: () => Promise<void>) {
+    if (this.destroyed) return
+    state.promise = fn()
+      .catch((err) => {
+        dbLogger.error({ err }, 'blob divert failed')
+      })
+      .finally(() => {
+        state.timer = setTimeout(() => this.poll(state, fn), 30 * SECOND)
+      })
+  }
+
+  async processAll() {
+    await Promise.all([this.divertBlob(), this.pollState.promise])
+  }
+
+  async destroy() {
+    this.destroyed = true
+    const destroyState = (state: PollState) => {
+      if (state.timer) {
+        clearTimeout(state.timer)
+      }
+      return state.promise
+    }
+    await destroyState(this.pollState)
+  }
+
+  async divertBlob() {
+    const toPush = await this.db.db
+      .selectFrom('blob_divert_event')
+      .select('id')
+      .forUpdate()
+      .skipLocked()
+      .where('divertedAt', 'is', null)
+      .where('attempts', '<', 10)
+      .execute()
+    await Promise.all(toPush.map((evt) => this.attemptBlobDivert(evt.id)))
+  }
+
+  private async getBlob({
+    pds,
+    did,
+    cid,
+  }: {
+    pds: string
+    did: string
+    cid: string
+  }) {
+    const blobResponse = await axios.get(
+      `${pds}/xrpc/com.atproto.sync.getBlob`,
+      {
+        params: { did, cid },
+        decompress: true,
+        responseType: 'stream',
+        timeout: 5000, // 5sec of inactivity on the connection
+      },
+    )
+    const imageStream: Readable = blobResponse.data
+    const verifyCid = new VerifyCidTransform(CID.parse(cid))
+    forwardStreamErrors(imageStream, verifyCid)
+
+    return {
+      contentType:
+        blobResponse.headers['content-type'] || 'application/octet-stream',
+      imageStream: imageStream.pipe(verifyCid),
+    }
+  }
+
+  private async uploadBlob(
+    {
+      imageStream,
+      contentType,
+    }: { imageStream: Readable; contentType: string },
+    { subjectDid, subjectUri }: { subjectDid: string; subjectUri: string },
+  ) {
+    if (!this.serviceConfig.authToken || !this.serviceConfig.url) {
+      return false
+    }
+
+    const url = `${this.serviceConfig.url}?did=${subjectDid}&uri=${subjectUri}`
+    const result = await axios(url, {
+      method: 'POST',
+      data: imageStream,
+      headers: {
+        Authorization: this.serviceConfig.authToken,
```

Comment:
Ideally we wouldn't need to configure the whole authorization header including the auth scheme.  Should we be using service auth here, a bearer token, or something else?  (CC @bnewbold)

Reply:
`abyss` wants HTTP Basic auth, with username `admin`, and the password as configured "token".

I don't think we have service-auth implemented generically in golang yet, so not super quick to add that (there are other golang clients for this service as well, so need both client+server impls). If we can keep it HTTP Basic that is best from my perspective. I don't know how the `axios` HTTP library works, maybe we could even encode the password in the URL so there is a single var to configure? `abyss` is generally not really set up as a regular atproto service, it is most like cardyb and other microservices (one diff is that abyss has pseudo-lexicon endpoints, though the golang side doesn't actually use the lexicon codegen or validation).